include/openthread/platform/toolchain.h

```diff
@@ -63,6 +63,22 @@
 extern "C" {
 #endif
 
+/**
+ * @def OT_MUST_USE_RESULT
+ *
+ * Compiler-specific indication that a class or enum must be used when it is
+ * the return value of a function.
+ *
+ * @note This is currently only available with clang (C++17 implements it
+ *       as attribute [[nodiscard]]).
+ *
+ */
+#if defined(__clang__)
+#define OT_MUST_USE_RESULT __attribute__((warn_unused_result))
```

Comment:
Can we allow OT API user to turn this feature off somehow? I think the major target of this PR is the OT core code.

Reply:
I don't see a strategy to make this happen, only if we add a new error class and use the new error class in OT core code. Do you have any suggestion? @simonlingoogle